### ISSUE TITLE
Keyboard Navigation Dialog

### ISSUE THREAD

**Issue by arouinfar** [phetsims/friction#45]
So I've been reviewing the latest dev ([1.4.0-dev.2](http://www.colorado.edu/physics/phet/dev/html/friction/1.4.0-dev.2/friction_en.html?accessibility)) and it seems like the book is somewhat slider-like, but without the page up/down and home/end shortcuts.

Here's my best guess for what content should be in the dialog. @terracoda can you review?

**Move Book** 
Move book: [arrow keys]
Move in smaller steps: [Shift] + [arrow keys]

**General Navigation**
Move to next item: [Tab]
Move to previous item: [Shift] + [Tab]
Exit this dialog: [Esc]

**Comment by terracoda** [comment]
Actually, the WASD keys work for moving the book as well, which makes me think we should have Up Arrow first. I'm just going to check RIAW's visual again.

**Comment by terracoda** [comment]
@arouinfar, I think we want to list the WASD keys as an alternative to the the Arrow keys, and it likely makes sense to have the arrow keys in the same order.

What do you thin of this?
**Move Book**
Move book: [Up Left Down Right], or
Move book:  [W  A  S  D] (up, left, down, or right) 
Move book in smaller steps: [Shift] + [Up Left Down Right], or
Move book in smaller steps: [Shift] + [W A S D]

General Navigation
Move to next item: [Tab]
Move to previous item: [Shift] + [Tab]
Exit this dialog: [Esc]

**Comment by arouinfar** [comment]
@terracoda the text looks good, but perhaps a little repetitive. I mocked up two options -- the first matches your text, and the second consolidates the Move Book section a bit. I chose to put the WASD keys first because the arrows follow the same order. 

![screen shot 2017-10-06 at 11 35 57 am](https://user-images.githubusercontent.com/8419308/31290677-967073da-aa8a-11e7-86dd-b5229e8c9041.png)
![screen shot 2017-10-06 at 11 36 10 am](https://user-images.githubusercontent.com/8419308/31290676-96373cb4-aa8a-11e7-9676-f93e3b9766b6.png)

**Comment by terracoda** [comment]
@arouinfar, I think the icon for the Arrow keys should go first. Users are likely more familiar with them. 

And if you line them up one above the other like you have it the second group in the second example, I think that will be nicely align up with W and so on.

**Move Book**
Move book: 
[Up Left Down Right keys icon] or
[W A S D keys icon] 

Move book in smaller steps: 
[Shift] + [Up Left Down Right keys icon] or
[Shift] + [W A S D keys icon]

@arouinfar, are you opposed to including the word, "or" in both cases? I think the "or" helps present the WASD keys as an alternative.

Happy to hear your thoughts, though.

**Comment by arouinfar** [comment]
>  I think the icon for the Arrow keys should go first. Users are likely more familiar with them.

Good point! I've swapped the order.

> @arouinfar, are you opposed to including the word, "or" in both cases? I think the "or" helps present the WASD keys as an alternative.

Not at all opposed to "or"! I left it out by mistake.

I think Shift + arrows/WASD needs to be split onto two lines because it's quite long, but I'm not sure it's necessary to split the up the move book keys too. I've mocked it up both ways, and have reduced the vertical padding between keys that wrap to the 2nd line so the "or" option looks more associated with the text above.

![screen shot 2017-10-06 at 12 03 31 pm](https://user-images.githubusercontent.com/8419308/31291848-6aee237a-aa8e-11e7-8703-11a1d438d2ec.png)
![screen shot 2017-10-06 at 12 03 18 pm](https://user-images.githubusercontent.com/8419308/31291847-6aea6d48-aa8e-11e7-8b03-d0b608977c1d.png)

Back to you @terracoda!

**Comment by terracoda** [comment]
@arouinfar, how do you feel about vertically left-aligning the move key icons? Like this?
![keyboard-dialog-2](https://user-images.githubusercontent.com/1902116/31292474-29cc177e-aaae-11e7-8346-e5af4ab9a2a8.png)

**Comment by arouinfar** [comment]
@terracoda I like your suggestion. If we were to left-align all of the move icons, I think we'd want to get rid of some of the empty space in the middle of the dialog, like this:
![image](https://user-images.githubusercontent.com/8419308/31292988-ca0d0d22-aa92-11e7-9c4b-c1cd4c59d716.png)

If we decide to stack the arrow keys and WASD keys in "Move Book", then I think it looks definitely better to left-align all of the move keys -- even though I do find it a little odd that "Move in smaller steps:" doesn't have any keys to its right.

However, I still really like the compactness of this layout:
![image](https://user-images.githubusercontent.com/8419308/31293128-5157f026-aa93-11e7-8d49-41d1fb995617.png)

**Comment by terracoda** [comment]
@arouinfar, I like the compactness of both layouts. The stacked version is compact in the other direction. I personally find the stacked version a little easier to read.

Shall we call in some tie breakers? @jessegreenberg, @zepumph, @mbarlow12, @emily-phet  which if the 2 layouts do you like best?

1. Stacked
2. Super compact

**Comment by emily-phet** [comment]
@terracoda @arouinfar +1 for Super compact.

**Comment by zepumph** [comment]
+1 super compact

**Comment by jessegreenberg** [comment]
+1 super compact

**Comment by terracoda** [comment]
OK, thanks for voting everyone.
@arouinfar, let's go with the Super Compact version!

@jessegreenberg, can you assign to the appropriate person for implementation?

**Comment by jessegreenberg** [comment]
@zepumph would you mind working on this at your convenience?  I added a few types to scenery-phet that should assist with layout but we might need something new to handle the "Move in smaller steps" portion of the layout.

**Comment by jessegreenberg** [comment]
We needed a way to add a list of controls in https://github.com/phetsims/resistance-in-a-wire/issues/104, so I added HelpContent.labelWithIconList for, hopefully it can be used here too!

**Comment by zepumph** [comment]
This is what master is looking like currently. Everything was pretty straight forward. Thanks @jessegreenberg for some awesome helper method/factory method work. The WASD keys are ugly, but I couldn't get the margins to work, and don't have current time to look into it. I just wanted to share progress. @jessegreenberg do you know why the xMargin and yMargin options might not work for me when creating KeyNodes in HelpContent.wasdRowIcon()? 

![image](https://user-images.githubusercontent.com/6856943/31696465-0ce4beaa-b35e-11e7-9a1e-86f35cf7b82f.png)

**Comment by jessegreenberg** [comment]
@zepumph to add padding arround the key icon, the current options are minKeyWidth and minKeyHeight. xMargin and yMargin are additional padding that can be used with xAlign and yAlign options to add additional padding for the layout. This is mostly useful if we need keys like this (this is what my escape key looks like)
![image](https://user-images.githubusercontent.com/6396244/31739323-badd822e-b40a-11e7-8948-f4a7deb86b16.png)

I believe that this will work for the letter keys, but let me know if you have concerns about the API.

**Comment by zepumph** [comment]
That works much nicer to me. Thanks! What do you think?

At some point it may be nice to factor out those "keyOptions" so that we have a general look of a single char KeyNode.

![image](https://user-images.githubusercontent.com/6856943/31743381-70457ca6-b407-11e7-9f21-5bbc2ed0d687.png)

**Comment by zepumph** [comment]
@terracoda, @emily-phet, @arouinfar what do you guys think about the dialog above? You can use it here, http://phettest.colorado.edu/friction/friction-a11y-view.html?brand=phet

**Comment by jessegreenberg** [comment]
Looks great, @zepumph thanks! I think it is important that the WASD keys are the same size as the arrow keys so that they vertically align.  Also, can you please align the letter text in the center of the key nodes?

**Comment by terracoda** [comment]
@zepumph, looks like you are making progress. I agree with @jessegreenberg that the Arrow keys and the WASD keys should be the same size. The WASD key letters need to be entered, too.

**Comment by zepumph** [comment]
@jessegreenberg would you please look over the implementation that I made. I think we should figure out a way to have all keys be the same size by default. Currently there are copied options in `ArrowKeyNode` and `HelpContent` where the wasd row is made. Perhaps these should go in KeyNode proper, but you would know best. I marked todo's so we wouldn't forget about it. @terracoda @jessegreenberg how is this:

![image](https://user-images.githubusercontent.com/6856943/32128659-25898ee2-bb2c-11e7-9b11-c743e6e2b30a.png)


It is the same amount of white margin as the arrow keys.

**Comment by zepumph** [comment]
Since they are hard coded values to get the two rows to be the same size, they may not be perfect. I would rather change the implementation that continue to pixel polished these hard coded values. Assigning to @jessegreenberg for discussion.

**Comment by terracoda** [comment]
@zepumph , this example looks much much better. The letters, words and arrows all see nicely centered with good white space. One small issue is that the drop shadows look a little uneven. This may be due to the hard-coded implementation.

Looking really good!

**Comment by arouinfar** [comment]
The dialog is looking much nicer @zepumph!! I also noticed the drop shadow issue @terracoda pointed out, which I think would be nice to address at some point.

**Comment by zepumph** [comment]
Key icon consistency will be addressed in https://github.com/phetsims/scenery-phet/issues/343, is there anything else for this issue @terracoda @arouinfar @jessegreenberg? After that issue Is this sim ready for republication with keyboard nav?

**Comment by jessegreenberg** [comment]
Looks great @zepumph!

> After that issue Is this sim ready for republication with keyboard nav?

I think so! But I would like to get @emily-phet opinion before proceeding with publication.

> One small issue is that the drop shadows look a little uneven.

@terracoda for reference, do you prefer the thicker drop shadow around the long keys or the thinner drop shadow around the letter/arrow keys?

**Comment by zepumph** [comment]
It should be noted that the wasd keys and arrows keys have different drop shadows.

**Comment by zepumph** [comment]
in phetsims/scenery-phet#343 we tackled a lot of the sizing issues and made things uniform. There may be a few pixel polishes that touch all of the Dialogs, but for the most part I think we are good to go.  Here is what it looks like now, it's also on master:
![image](https://user-images.githubusercontent.com/6856943/32684197-e05b099c-c62d-11e7-92f5-2c13bff27fb2.png)

@arouinfar @terracoda how does the drop shadow look to you?

**Comment by terracoda** [comment]
@zepumph, the drop shadows look good to me. Nice work!

The white space within the keys, however, seems to have changed. For example, the W key is wider than the other letter keys, and the arrow keys have a lot less white space around the arrows than before. 

Drop shadow is good! You might be working on the white space separately.

**Comment by arouinfar** [comment]
The drop shadows look great @zepumph! As @terracoda pointed out, the white space within the keys looks a bit off. All arrow and letter keys should be equally sized squares.

**Comment by zepumph** [comment]
Thanks for the feedback. 

I haven't yet centralized how to make keys squares. That is next on my list. I wonder if we should make it so that the sqaureKey option forces the margin down, or it scales down the node. It seems like it should scale down the node, but that looks the worst, so I think I want to do a combo solution where it scales down the node if it doesn't have the margin to allow for it.

One other solution is that we just sorta picked that the height is 15, arbitrarily. Maybe if we made the height a bit bigger, then the "square" functionality would just happen by default for the larger keys like 'W' and 'M'. Another thought is that we could use a smaller PhetFont (currently 12).

**Comment by zepumph** [comment]
@jessegreenberg is this now ready to go?

**Comment by jessegreenberg** [comment]
It looks great @zepumph thanks, ready to close this issue.

**Comment by jessegreenberg** [comment]
Reopening since we are going to change the look after https://github.com/phetsims/a11y-research/issues/62.

**Comment by jessegreenberg** [comment]
Here is the updated content in Friction:
![friction](https://user-images.githubusercontent.com/6396244/33100178-b1cb95a0-ced0-11e7-9827-5b0c27194b58.PNG)

**Comment by zepumph** [comment]
Looks great thanks!

**Comment by zepumph** [comment]
Ready to close?

**Comment by jessegreenberg** [comment]
I think so! We got one 👍 in https://github.com/phetsims/a11y-research/issues/62.
